tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+	dryRun               = flag.Bool("dry-run", false, "Performs a dry run for all deletion functions.")
+	concurrentOperations = flag.Int("concurrent-operations", 10, "How many deletion operations to run concurrently.")
+
+	// Authentication method for using Google Cloud Registry APIs.
+	auther = authn.DefaultKeychain
+
+	// Client for deleting GKE clusters.
+	gkeClient gke.SDKOperations
+)
+
+// selectProjects returns the list of projects to iterate over.
+func selectProjects(project, resourceFile, regex string) ([]string, error) {
+	// --project used, just return it.
+	if project != "" {
+		log.Printf("Iterating over projects [%s]", project)
+		return []string{project}, nil
+	}
+	// Otherwise, read the resource file and extract the project names.
+	projectRegex, err := regexp.Compile(regex)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid regular expression %q", regex)
+	}
+	content, err := ioutil.ReadFile(resourceFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot read file %q", resourceFile)
+	}
+	projects := make([]string, 0)
+	for _, line := range strings.Split(string(content), "\n") {
+		if len(line) > 0 {
+			if p := projectRegex.Find([]byte(line)); p != nil {
+				projects = append(projects, string(p))
+			}
+		}
+	}
+	if len(projects) == 0 {
+		return nil, fmt.Errorf("no project found in %q matching %q", resourceFile, regex)
+	}
+	log.Printf("Iterating over projects defined in %q, matching %q", resourceFile, regex)
+	return projects, nil
+}
+
+// deleteImage deletes a single GCR image pointed by the given reference.
+func deleteImage(ref string) error {
+	image, err := name.ParseReference(ref)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse reference %q", ref)
+	}
+	if err := remote.Delete(image, remote.WithAuthFromKeychain(auther)); err != nil {
+		return errors.Wrapf(err, "failed to delete %s", image)
+	}
+	return nil
+}
+
+// deleteClusters deletes old clusters from a given project.
+func deleteClusters(project string, before time.Time) (int, error) {
+	// TODO(adrcunha): Consider exposing https://github.com/knative/pkg/blob/6d806b998379948bd0107d77bcd831e2bdb4f3cb/testutils/clustermanager/e2e-tests/gke.go#L281
+	if project == "knative-tests" {
+		return 0, fmt.Errorf("cleaning up %q is forbidden", project)
+	}
+	// List clusters, delete those created before the given timestamp.
+	clusters, err := gkeClient.ListClustersInProject(project)
+	if err != nil {
+		return 0, errors.Wrap(err, "error listing clusters in %q, maybe try 'gcloud auth application-default login'")
+	}
+	count := 0
+	for _, cluster := range clusters {
+		creation, err := time.Parse(time.RFC3339, cluster.CreateTime)
+		if err != nil {
+			return count, errors.Wrapf(err, "error getting creation time for cluster %q", cluster.Name)
+		}
+		age := int(time.Since(creation).Hours())
+		log.Printf("%s/%s is %d hours old", project, cluster.Name, age)
+		if creation.Before(before) {
+			if *dryRun {
```

Comment:
this dryRun logic can be refactored out or maybe use this one: https://github.com/knative/pkg/blob/master/test/helpers/dryrun.go

Reply:
Done.